File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
#         my_set = set()
        
#         for num in nums :
#             if num in my_set:
#                 return num
#             else :
#                 my_set.add(num)
            
    
           # for mutable array
           # for i in range(0,len(nums)) :
           #      if nums[abs(nums[i])] < 0 :
           #          return abs(nums[i])
           #      else :
           #          nums[abs(nums[i])] = -nums[abs(nums[i])]
            
            slow , fast = nums[0] , nums[0]
            while True :
                slow = nums[slow]
                fast = nums[nums[fast]]
                if slow == fast :
                    break
            
            fast = nums[0]
            
            while slow != fast :
                slow = nums[slow]
                fast = nums[fast]
                
            return fast
            

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        HashMap<Integer,Boolean> mpp = new HashMap<>();
        
        for(int i=0;i<nums.length;i++){
            if(mpp.containsKey(nums[i]) == true){
                return nums[i];
            }
            mpp.put(nums[i],true);
        }
        
        return 0;
    }
}